rita/src/rita_common/dao_manager/mod.rs

```diff
@@ -0,0 +1,278 @@
+    }
+}
+
+impl DAOManager {
+    fn new() -> DAOManager {
+        DAOManager {
+            entries: HashMap::<Identity, Vec<DAOEntry>>::new(),
+        }
+    }
+}
+
+pub struct DAOEntry {
+    on_list: bool,
+    dao_address: EthAddress,
+    id: Identity,
+    last_updated: Instant,
+}
+
+pub struct DAOCheck(Identity);
+impl Message for DAOCheck {
+    type Result = ();
+}
+
+impl Handler<DAOCheck> for DAOManager {
+    type Result = ();
+
+    fn handle(&mut self, msg: DAOCheck, _: &mut Context<Self>) -> Self::Result {
+        let their_id = msg.0;
+        check_cache(their_id, &self.entries);
+    }
+}
+
+/// Called by returning DAO requests, sends a message to TunnelManager
+pub struct CacheCallback((Identity, EthAddress, Web3Response));
+impl Message for CacheCallback {
+    type Result = ();
+}
+
+impl Handler<CacheCallback> for DAOManager {
+    type Result = ();
+
+    fn handle(&mut self, msg: CacheCallback, _: &mut Context<Self>) -> Self::Result {
+        let their_id = (msg.0).0;
+        let dao_address = (msg.0).1;
+        let response = (msg.0).2;
+
+        let has_vec = self.entries.contains_key(&their_id);
+        let on_dao = response.result == Uint256::zero();
+
+        if has_vec {
+            let entry_vec = self.entries.get_mut(&their_id).unwrap();
+            let mut found_entry = false;
+            for entry in entry_vec.iter_mut() {
+                // There exists an entry, and we found it, so update it
+                if dao_address == entry.dao_address {
+                    entry.on_list = on_dao;
+                    entry.last_updated = Instant::now();
+                    found_entry = true;
+                    send_membership_message(on_dao, their_id.clone());
+                    break;
+                }
+            }
+            if !found_entry {
+                // A list exists but does not contain an entry from this dao
+                // create one and insert it into the list
+                let entry = DAOEntry {
+                    on_list: on_dao,
+                    dao_address: dao_address,
+                    id: their_id.clone(),
+                    last_updated: Instant::now(),
+                };
+                entry_vec.push(entry);
+                send_membership_message(on_dao, their_id);
+            }
+        } else {
+            // No entry exists in the HashMap for this ID, create one
+            let entry = DAOEntry {
+                on_list: on_dao,
+                dao_address: dao_address,
+                id: their_id.clone(),
+                last_updated: Instant::now(),
+            };
+            let mut v = Vec::new();
+            v.push(entry);
+            self.entries.insert(their_id.clone(), v);
+
+            send_membership_message(on_dao, their_id);
+        }
+    }
+}
+
+/// True if timestamp does not need to be updated
+fn timer_check(timestamp: Instant) -> bool {
+    if Instant::now() - timestamp < SETTING.get_dao().cache_timeout {
```

Comment:
You could drop the if/else and just return the result of the condition here

Reply:
+1